### DISCUSSION TITLE
Sequence diagrams - support for options, alternatives, loops?

### DISCUSSION THREAD

**Discussion by AndreiPashkin** [terrastruct/d2#1688] [Ideas]
UML sequence diagrams are supposed to support option, alternative and loop fragments:
![image](https://github.com/terrastruct/d2/assets/4378647/5a9be9c4-cea8-4776-8efc-11a62e637ddd)
![image](https://github.com/terrastruct/d2/assets/4378647/c08041e7-434a-422f-990a-a41eb06fe0ff)

But there is no explicit support for this in D2. The only way you could reflect those constructs right now is like this:
```
shape: sequence_diagram
# Predefine actors
alice
bob
alt: {
  condition1: {
    alice -> bob: Action 1
  }
  condition2: {
    alice -> bob: Action 2
  }
  else: {
    alice -> bob: Action 3
  }
}
opt: {
  condition4: {
    bob -> alice: Action 4
  }
}
loop: {
  condition5: {
    bob -> alice: Action 5
  }
}
```
![d2](https://github.com/terrastruct/d2/assets/4378647/912cfa1b-90ed-4993-9eb9-aa078436ae43)

Should an explicit support for these things be added?

**Comment by alixander** [comment]
I don't understand what explicit support means.

It looks like the difference between "option, alternative and loop fragments" are just the labels, and you can use any label in D2.

**Reply by terlar** [reply]
I agree for the simple case showcased here, what about alts that use several sections and have secondary labels, such as:
![image](https://github.com/terrastruct/d2/assets/280235/e0301742-d367-4b96-83b0-dd6679f4f2cf)

**Reply by terlar** [reply]
There is also dividers:
![image](https://github.com/terrastruct/d2/assets/280235/01723951-62d5-471d-9af3-9cb1646b5acf)

And delays:
![image](https://github.com/terrastruct/d2/assets/280235/1e6bdc3d-f630-4974-af24-b47faec9af79)

Is it possible to achieve somehow?

**Reply by sanguineti** [reply]
I believe the best approach is for D2 to natively support all the features of a sequence diagram according to the UML standard. As @terlar  evidenced above, it is not just a label.

**Reply by alixander** [reply]
@terlar Thanks for the examples. I'm convinced D2 needs something akin to the alts, delays, and dividers, as well as secondary labels like the `[successful case]` next to `alt`.

Can you provide an isolated example of loop and option being necessary?

**Reply by sanguineti** [reply]
@alixander Here are some examples

![image](https://github.com/terrastruct/d2/assets/44387191/8b0e0200-c8f0-4c8c-bd6b-cd47a52b420e)
![image](https://github.com/terrastruct/d2/assets/44387191/ff01ceaa-4b50-4c9d-a42d-f37469d4ce00)

**Comment by OPetrusBet** [comment]
Hi. Any update on that? Looks like this feature is still missing.
